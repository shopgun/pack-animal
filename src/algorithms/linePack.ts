import { IPolygon, polygonArea, polygonHeight, polygonWidth } from "../geometry";
import { average } from "../maths";
import { getPolygonTransform, ITransform } from "../transform";
import { noop } from "../utilities";
import { Matrix } from "../vendor/matrix";

export const linePack = (
  rectangleWidth: number,
  rectangleHeight: number,
  polygonList: IPolygon[],
  vertical: boolean = false,
  { debug: dbug = noop, averageArea: averageAreaOption = 0, alignBottom = true, margin = 0 } = {}
): ITransform[] => {
  const polygons = polygonList.map(({ points }) => points);
  // Wrap debug function to include current algorithm.
  const debug = (...args: any[]) => dbug("linePack:", ...args);
  // Write out said algorithm entry.
  debug();

  const polygonAreas = polygons.map(polygon => polygonArea(polygon));
  const averageArea = averageAreaOption || average(polygonAreas);
  const polygonNormalizingScales = polygons.map(
    (_, i) => (averageArea / polygonAreas[i] - 1) / 2 + 1
  );
  const polygonWidths = polygons.map(
    (polygon, i) => polygonWidth(polygon) * polygonNormalizingScales[i] + margin
  );
  const polygonHeights = polygons.map(
    (polygon, i) => polygonHeight(polygon) * polygonNormalizingScales[i]
  );

  if (vertical) {
    const maxWidth = Math.max(
      ...polygons.map((_, i) => polygonWidths[i] * polygonNormalizingScales[i])
    );

    let verticalOffset = 0;
    return polygons.map((polygon, i) => {
      const normalizingScale = polygonNormalizingScales[i];
      const horizontalCorrection = maxWidth - polygonWidths[i] / 2;
      const matrix = new Matrix()
        .translate(horizontalCorrection, verticalOffset)
        .scaleU(normalizingScale);
      const polygonTransform = getPolygonTransform(
        rectangleWidth,
        rectangleHeight,
        polygon,
        matrix
      );
      verticalOffset += polygonHeights[i];
      return polygonTransform;
    });
  } else {
    const maxHeight = Math.max(
      ...polygons.map((_, i) => polygonHeights[i] * polygonNormalizingScales[i])
    );

    let horizontalOffset = 0;
    return polygons.map((polygon, i) => {
      const normalizingScale = polygonNormalizingScales[i];
      const verticalCorrection = alignBottom
        ? maxHeight - polygonHeights[i]
        : maxHeight - polygonHeights[i] / 2;
      const matrix = new Matrix()
        .translate(horizontalOffset, verticalCorrection)
        .scaleU(normalizingScale);
      const polygonTransform = getPolygonTransform(
        rectangleWidth,
        rectangleHeight,
        polygon,
        matrix
      );
      horizontalOffset += polygonWidths[i];
      return polygonTransform;
    });
  }
};
