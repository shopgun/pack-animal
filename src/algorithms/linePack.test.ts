import { linePack } from "./linePack";

describe("linePack", () => {
  it("works", () => {
    const polygonTransforms = linePack(
      800,
      742,
      [[{ x: 1, y: 1 }, { x: 1, y: 366 }, { x: 798, y: 366 }, { x: 798, y: 1 }]].map(
        (points, index) => ({ points, index })
      )
    );
    expect(polygonTransforms).toMatchSnapshot();

    const polygonTransforms2 = linePack(
      720,
      300,
      [
        [{ x: 0, y: 0 }, { x: 138, y: 0 }, { x: 138, y: 324 }, { x: 0, y: 324 }, { x: 0, y: 0 }],
        [{ x: 0, y: 0 }, { x: 163, y: 0 }, { x: 163, y: 349 }, { x: 0, y: 349 }, { x: 0, y: 0 }],
        [{ x: 0, y: 0 }, { x: 148, y: 0 }, { x: 148, y: 364 }, { x: 0, y: 364 }, { x: 0, y: 0 }],
        [{ x: 0, y: 0 }, { x: 150, y: 0 }, { x: 150, y: 350 }, { x: 0, y: 350 }, { x: 0, y: 0 }]
      ].map((points, index) => ({ points, index })),
      720 > 300,
      { margin: 12 }
    );
    expect(polygonTransforms2).toMatchSnapshot();
  });
});
