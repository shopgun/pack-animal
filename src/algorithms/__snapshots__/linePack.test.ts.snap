// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`linePack works 1`] = `
Array [
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 99.625%;
        height: 49.19137466307278%;
        z-index: 3650000000797;
        transform: translateX(0%) translateY(0%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":1,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":1,
      \\"e\\":0,
      \\"f\\":0
    }",
    "points": Array [
      Object {
        "x": 1,
        "y": 1,
      },
      Object {
        "x": 1,
        "y": 366,
      },
      Object {
        "x": 798,
        "y": 366,
      },
      Object {
        "x": 798,
        "y": 1,
      },
    ],
    "rotate": -0,
    "scale": 1,
    "svgTransform": "matrix(1,0,0,1,0,0)",
    "translateX": 0,
    "translateY": 0,
    "zIndex": 3650000000797,
  },
]
`;

exports[`linePack works 2`] = `
Array [
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 20.727184070644718%;
        height: 116.79317632850241%;
        z-index: 350;
        transform: translateX(62.81699707465993%) translateY(0%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":1.0814182993379853,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":1.0814182993379853,
      \\"e\\":93.74540120147716,
      \\"f\\":0
    }",
    "points": Array [
      Object {
        "x": 93.74540120147716,
        "y": 0,
      },
      Object {
        "x": 242.98112651011914,
        "y": 0,
      },
      Object {
        "x": 242.98112651011914,
        "y": 350.37952898550725,
      },
      Object {
        "x": 93.74540120147716,
        "y": 350.37952898550725,
      },
      Object {
        "x": 93.74540120147716,
        "y": 0,
      },
    ],
    "rotate": -0,
    "scale": 1.0814182993379853,
    "svgTransform": "matrix(1.0814182993379853,0,0,1.0814182993379853,93.74540120147716,0)",
    "translateX": 93.74540120147716,
    "translateY": 0,
    "zIndex": 350,
  },
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 21.66502507163324%;
        height: 111.32898773006134%;
        z-index: 684;
        transform: translateX(57.93334680815037%) translateY(104.90814540745671%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":0.9569827025506706,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":0.9569827025506706,
      \\"e\\":90.36917359791849,
      \\"f\\":350.37952898550725
    }",
    "points": Array [
      Object {
        "x": 90.36917359791849,
        "y": 350.37952898550725,
      },
      Object {
        "x": 246.35735411367781,
        "y": 350.37952898550725,
      },
      Object {
        "x": 246.35735411367781,
        "y": 684.3664921756913,
      },
      Object {
        "x": 90.36917359791849,
        "y": 684.3664921756913,
      },
      Object {
        "x": 90.36917359791849,
        "y": 350.37952898550725,
      },
    ],
    "rotate": -0,
    "scale": 0.9569827025506706,
    "svgTransform": "matrix(0.9569827025506706,0,0,0.9569827025506706,90.36917359791849,350.37952898550725)",
    "translateX": 90.36917359791849,
    "translateY": 350.37952898550725,
    "zIndex": 684,
  },
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 20.197029532967033%;
        height: 119.21706081081082%;
        z-index: 1042;
        transform: translateX(65.77834556547015%) translateY(191.3502669056552%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":0.9825581934956935,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":0.9825581934956935,
      \\"e\\":95.65395753711684,
      \\"f\\":684.3664921756913
    }",
    "points": Array [
      Object {
        "x": 95.65395753711684,
        "y": 684.3664921756913,
      },
      Object {
        "x": 241.07257017447947,
        "y": 684.3664921756913,
      },
      Object {
        "x": 241.07257017447947,
        "y": 1042.0176746081238,
      },
      Object {
        "x": 95.65395753711684,
        "y": 1042.0176746081238,
      },
      Object {
        "x": 95.65395753711684,
        "y": 684.3664921756913,
      },
    ],
    "rotate": -0,
    "scale": 0.9825581934956935,
    "svgTransform": "matrix(0.9825581934956935,0,0,0.9825581934956935,95.65395753711684,684.3664921756913)",
    "translateX": 95.65395753711684,
    "translateY": 684.3664921756913,
    "zIndex": 1042,
  },
  Object {
    "cssText": "
        position: absolute;
        top: 0;
        left: 0;
        transform-origin: 0 0;
        width: 20.732688492063495%;
        height: 116.10305555555556%;
        z-index: 1390;
        transform: translateX(62.787044746222776%) translateY(299.1645854687881%) rotate(0deg)
      ",
    "matrix": "{
      \\"a\\":0.9951690476190476,
      \\"b\\":0,
      \\"c\\":0,
      \\"d\\":0.9951690476190476,
      \\"e\\":93.72558528436959,
      \\"f\\":1042.0176746081238
    }",
    "points": Array [
      Object {
        "x": 93.72558528436959,
        "y": 1042.0176746081238,
      },
      Object {
        "x": 243.00094242722673,
        "y": 1042.0176746081238,
      },
      Object {
        "x": 243.00094242722673,
        "y": 1390.3268412747905,
      },
      Object {
        "x": 93.72558528436959,
        "y": 1390.3268412747905,
      },
      Object {
        "x": 93.72558528436959,
        "y": 1042.0176746081238,
      },
    ],
    "rotate": -0,
    "scale": 0.9951690476190476,
    "svgTransform": "matrix(0.9951690476190476,0,0,0.9951690476190476,93.72558528436959,1042.0176746081238)",
    "translateX": 93.72558528436959,
    "translateY": 1042.0176746081238,
    "zIndex": 1390,
  },
]
`;
